### PR TITLE
doc: add Kubernetes platform-optimization README

### PR DESCRIPTION
**Pull request by askervin**
Add documentation that helps optimizing Kubernetes nodes for OPEA AI applications.

### CODE REVIEW

**Review by poussa — APPROVED** [review]
LGTM

**Review by kding1 — APPROVED** [review]
LGTM.

File: doc/platform-optimization/README.md [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+In this example, we use the balloons policy because it can be tuned
+for certain applications (like RAG pipelines) using even node-specific
+parameters for each container in applications. The topology-aware
+policy, on the other hand, needs no configuration and does CPU
+assignment automatically based on resource requests in containers and
+underlying hardware topology.
+
+## Install
+
+Warning: installing and reconfiguring the balloons policy in can
```
Comment (mkbhanda) [inline doc/platform-optimization/README.md:45]: s/in/

Also how can one tell if there are containers running that are relying on some static allocation?

File: doc/platform-optimization/README.md [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+Install the balloons policy with helm:
+
+1. Add the NRI plugins repository
+   ```bash
+   helm repo add nri-plugins https://containers.github.io/nri-plugins
+   ```
+
+2. Install the balloons resource policy and patch container runtime's
+   configuration on host in order to switch NRI support on.
```
Comment (mkbhanda) [inline doc/platform-optimization/README.md:58]: s/configuration on host in order to switch NRI support on./configuration on the individual worker nodes/hosts to enable NRI support.

File: doc/platform-optimization/README.md [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+Let us consider isolating AI inference and reranking containers in
+[ChatQnA](https://github.com/opea-project/GenAIExamples/tree/main/ChatQnA)
+application's Gaudi accelerated pipeline.
+
+In the
+[manifest](https://github.com/opea-project/GenAIExamples/blob/main/ChatQnA/kubernetes/manifests/gaudi/chatqna.yaml)
+there are "tgi" and "tei" containers that will need a lot of CPUs.
+
+We will allocate 16 CPUs (8 full CPU cores with two hyperthreads on
```
Comment (mkbhanda) [inline doc/platform-optimization/README.md:83]: May want to add a description sentence like "We refer to physical cores as CPU Cores or full CPU cores and hyperthreaded cores as virtual cores or vCPUs or just CPUs".

.. (16 physical cores that are hyperthreaded making for 32 vCPUs)


Sometimes we need to turn off hyperthreading ... NRI can detect that.

File: doc/platform-optimization/README.md [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+  a balloon. Created balloon will never be smaller even if containers
+  assigned to a balloon of this type would request fewer or no CPUs at
+  all. Correspondingly `maxCPUs` could be used to set an upper limit
+  for CPUs.
+- `hideHyperthreads: true` means that containers in balloons of this
+  type are allowed to use only single CPU hyperthread from each full
+  CPU core in the balloon. By default, both using hyperthreads of all
+  CPUs in the balloon is allowed. Note that when `true`, both
+  hyperthreads are allocated to the balloon in any case, preventing
+  alloating them into other balloons. This ensures that the whole CPU
```
Comment (mkbhanda) [inline doc/platform-optimization/README.md:174]: s/alloating/allocating

File: doc/platform-optimization/README.md [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+  name: default
+spec:
+  allocatorTopologyBalancing: true
+  balloonTypes:
+  - name: tgi
+    allocatorPriority: high
+    minCPUs: 16
+    minBalloons: 1
+    preferNewBalloons: true
+    hideHyperthreads: true
```
Comment (mkbhanda) [inline doc/platform-optimization/README.md:101]: Line 118 says hidehyperthreads false and here true. May need to add a sentence or two on why this different balloon policy setting for tgi versus tei based on the application characteristics/behavior.

File: doc/platform-optimization/README.md [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+```
+
+The balloons policy creates "balloons" of CPUs that only containers
+assigned into a balloon are allowed to use.
+
+The most important options in the above configuration example are:
+
+- `allocatorTopologyBalancing: true`. This option ensures that
+  balloons (sets of allowed CPUs) are balanced between CPU sockets in
+  the system, and local NUMA nodes if the system is running a sub-NUMA
```
Comment (mkbhanda) [inline doc/platform-optimization/README.md:146]: May need to provide the reader a link to better understand what is NUMA and sub-NUMA and

File: doc/platform-optimization/README.md [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+  clustering (SNC) mode. Without this options balloons would be
+  tightly packed on a single socket allowing the other CPU socket
+  sleep and save power. Here we have optimized for performance.
```
Comment (mkbhanda) [inline doc/platform-optimization/README.md:149]: s/sleep/to sleep.

Add a sentence, "To optimize for power savings, one could alternately have set allocatorTopologyBalancing: false"

File: doc/platform-optimization/README.md [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+  containers into their special balloon types based on container
+  name. Matching could be done based on labels and pod name, too.
+- `preferNewBalloon: true` on both tei and tgi balloon types means
+  that when a container is assigned into this balloon type and it is
+  possible to create a new balloon of this type because there are
+  enough free CPUs in the system, then the new balloon will be created
+  for the container. As a result, both tei and tgi containers will get
+  dedicated set of CPUs, unlike other containers that will run in the
+  default balloon type. Each container is allowed to use only CPUs of
+  the balloon where they are assigned.
```
Comment (mkbhanda) [inline doc/platform-optimization/README.md:163]: With new, I am assuming from your description that a given CPU, physical or hyperthreaded is allocated to only one balloon.

is there a performance penalty when using a balloon for a workload that spans NUMA boundaries .. all the cache and memory reference related added latency?

File: doc/platform-optimization/README.md [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+  alloating them into other balloons. This ensures that the whole CPU
+  core is dedicated to containers in these balloons only.
+
+For more information about the configuration and the balloons resource
+policy, refer to the balloons
+[documentation](https://containers.github.io/nri-plugins/stable/docs/resource-policy/policy/balloons.html).
+
+## NRI topology-aware resource policy
+
+NRI plugins include topology-aware resource policy, too. Unlike
```
Comment (mkbhanda) [inline doc/platform-optimization/README.md:183]: Looks like using topology aware resource policy is easier .. might be good to add a sentence or two stating why take on the extra effort of balloons such as greater control, performance improvements of x% above and beyond ..

**Review by mkbhanda — COMMENTED** [review]
Minor edits and some clarifying questions. Thank you for an excellent document.

File: doc/platform-optimization/README.md (hunk 1 above)
Comment (askervin) [reply doc/platform-optimization/README.md:45]: It's impossible to automatically detect all containers that need static allowed CPUs, though there are some signs (for instance application setting CPU affinity of its own processes/threads) that suggest so.

Currently there is no generally agreed way to annotate a container with a "I can't handle my allowed CPUs changing". Perhaps this kind of annotation would enable runtime to restart/recreate containers whose CPUs need to be changed.

File: doc/platform-optimization/README.md (hunk 2 above)
Comment (askervin) [reply doc/platform-optimization/README.md:58]: Thanks, fixed!

File: doc/platform-optimization/README.md (hunk 3 above)
Comment (askervin) [reply doc/platform-optimization/README.md:83]: Thanks, added!

File: doc/platform-optimization/README.md (hunk 5 above)
Comment (askervin) [reply doc/platform-optimization/README.md:101]: Parameters of balloon type definitions written in a bit different order (always starting with name) to make it easier to notice that hideHyperthreads: false is defined for the default balloon only.

I also added explanation of hideHyperthreads: false separately. There it was also worth mentioning that when hyperthreading is off in BIOS, hideHyperthreads: false does not unhide them.

File: doc/platform-optimization/README.md (hunk 6 above)
Comment (askervin) [reply doc/platform-optimization/README.md:146]: Link added.

File: doc/platform-optimization/README.md (hunk 7 above)
Comment (askervin) [reply doc/platform-optimization/README.md:149]: Thank you, added!

File: doc/platform-optimization/README.md (hunk 8 above)
Comment (askervin) [reply doc/platform-optimization/README.md:163]: Your assumption is correct. I also added a clarification on that: a balloon can belong to at most one CPU at a time.

Spanning over NUMA boundaries is quite likely to hurt performance, especially if NUMA nodes are in different sockets. Sometimes it's better to have fewer CPUs and all data in local memory than more CPUs and a lot of memory accesses to remote NUMA.

File: doc/platform-optimization/README.md (hunk 4 above)
Comment (askervin) [reply doc/platform-optimization/README.md:174]: Thanks, fixed!

File: doc/platform-optimization/README.md (hunk 9 above)
Comment (askervin) [reply doc/platform-optimization/README.md:183]: Added few sentences. However, did not give performance numbers compared to topology-aware policy, as they vary quite a bit based on resource requests/limits of containers. The topology-aware uses that information for coming up with cpusets. Currently our example pipeline does not include any container resource information, so the policy cannot do much.

**Pull request merged**